### ISSUE TITLE
Sympa caching

### ISSUE THREAD

**Issue by dverdin** [sympa-community/sympa#584]
Here is an issue where we can put together ideas about how to improve the way Sympa caches informations.

**Comment by salaun-urennes1** [comment]
Soji invited me to check this issue, since I proposed in  #925 to use List.pm caching mecanism to solve the performance problem. Before I add a PR with more caching, I prefer not to if it's not the way to go, for some reason...

I read #583 and David's patch, and then I noticed Stefan and Soji's ploan to get rid of Sympa caching mecanisms.

@soji, @racke could you provide more details about you plans for Sympa caching replacement? What issues did you encounter with current caching? What technology do you plan to use?

Thanks

**Comment by racke** [comment]
On 5/7/20 9:03 AM, salaun-urennes1 wrote:
> Soji invited me to check this issue, since I proposed in  #925 <https://github.com/sympa-community/sympa/pull/925> to
> use List.pm caching mecanism to solve the performance problem. Before I add a PR with more caching, I prefer not to if
> it's not the way to go, for some reason...
> 
> I read #583 <https://github.com/sympa-community/sympa/pull/583> and David's patch, and then I noticed Stefan and Soji's
> ploan to get rid of Sympa caching mecanisms.
> 
> @Soji <https://github.com/Soji>, @racke <https://github.com/racke> could you provide more details about you plans for
> Sympa caching replacement? What issues did you encounter with current caching? What technology do you plan to use?

Well, we learned that the caching introduced before 6.2.16 caused lot of troubles *and* was done with homegrown code not
using some established caching framework (e.g. Cache::CHI).

Caching makes only sense for presentation of data (which can get stale anyway when the user goes for coffee break or
so), and not for transactions (changing user role, delete moderator, process incoming email, evaluate scenarios, ...)
which is the core business of Sympa.

So from my point of the view, heavy caching makes only sense when you move stuff out of the current Sympa code and
replace it with API calls to a API server which of course doesn't exist.

Regards
        Racke



-- 
Ecommerce and Linux consulting + Perl and web application programming.
Debian and Sympa administration. Provisioning with Ansible.

**Comment by salaun-urennes1** [comment]
> Well, we learned that the caching introduced before 6.2.16 caused lot of troubles *and* was done with homegrown code not using some established caching framework (e.g. Cache::CHI). 
Caching is done with homegrown code because it was developed many years ago.
I knew about memcache, but not Cache::CHI. How do the compare?

> Caching makes only sense for presentation of data (which can get stale anyway when the user goes for coffee break or so), and not for transactions (changing user role, delete moderator, process incoming email, evaluate scenarios, ...) which is the core business of Sympa. 
I agree that the main use of caching is to provide good performances on the GUI. However other Sympa interfaces (the CLI for instance) may also benefit from caching. 

> So from my point of the view, heavy caching makes only sense when you move stuff out of the current Sympa code and replace it with API calls to a API server which of course doesn't exist. 
By "API calls" you mean having a central Sympa web service, other Sympa services being clients of that web service?
Is this a short term plan?

**Comment by racke** [comment]
Yes, and it is certainly not a short term plan given the current capacities.

**Comment by salaun-urennes1** [comment]
But a new caching mechanism would make sense, even if we don't have a central Sympa web service in place...

**Comment by racke** [comment]
It is of limited use though as we can't use it for transactions. Reducing SQL queries during the lifetime of a request (HTTP or task) might be more efficient. 

And we would need to make sure that it doesn't screw things up again.

**Comment by ikedas** [comment]
> @Soji, @racke could you provide more details about you plans for Sympa caching replacement? What issues did you encounter with current caching? What technology do you plan to use?

I think that the cache mechanism of Sympa so far essentially did not make GUI response faster.
- Since the most of accesses by users are sparse, cached data won't improve performance in the most of cases.  Rather, especially on the first access, performance will _never_ be improved.
- Also, cache algorithm _cannot_ be as smart as prefethcing data to which the user probablly will access.
- By these reasons, the cache began to show its effect after slightly long time, i.e. when it has cached large amount of data.

Indeed homegrown code had several problems. However, either appropriately corrected code or established cache framework, cache may _not_ overcome problems above: Cache will be less effective, or will show effect slowly.

That's why I support deprecation of caching.

----
By the way, there is a  promising solution: _Cache everything_ at startup.  By this, cache miss by inappropriate algorithm will be entirely avoided, and cache will show effect just after startup.

However, it is _no longer the "cache"_: It's a verbatim copy of storage (database or filesystem) on the memory.  Several sites using Sympa support tens of thousands of list with hundreds of thousands of users or more.  Caching everithing is not scalable solution.

**Comment by salaun-urennes1** [comment]
> I think that the cache mechanism of Sympa so far essentially did not make GUI response faster.
> 
>     * Since the most of accesses by users are sparse, cached data won't improve performance in the most of cases.  Rather, especially on the first access, performance will _never_ be improved.

I can see two potential benefits for cached data: 1) user A initiates cache and user B benefits from it later on, but also 2) cache gets initiated at the beginning of user A session and prevents data reloading within that user session.  => That's what I would like to introduce as an alternative to #925. Running a single SQL query early and populate the cache to prevent one SQL query per list, to fulfill that request.

>     * Also, cache algorithm _cannot_ be as smart as prefetching data to which the user probablly will access.
> 
>     * By these reasons, the cache began to show its effect after slightly long time, i.e. when it has cached large amount of data.
> 
> 
> Indeed homegrown code had several problems. However, either appropriately corrected code or established cache framework, cache may _not_ overcome problems above: Cache will be less effective, or will show effect slowly.
> 
> That's why I support deprecation of caching.
> 
> By the way, there is a promising solution: _Cache everything_ at startup. By this, cache miss by inappropriate algorithm will be entirely avoided, and cache will show effect just after startup.
> 
> However, it is _no longer the "cache"_: It's a verbatim copy of storage (database or filesystem) on the memory. Several sites using Sympa support tens of thousands of lists with hundreds of thousands of users or more. Caching everything is not scalable solution.

Thanks for explaining your view about caching.

> EDIT:
> 
> N.B.: Personally, I've given up improving web UI with current codebase. Moreover, I think web UI of Sympa would be better to be rewritten from scratch. One reason is that all of security advisories we have published in the past are regarding web UI.

From time to time, if you can move most of business logic out of wwsympa.fcgi in the libraries, you could end up with a web GUI that's easier to rewrite...

**Comment by racke** [comment]
The problem is that the libraries are too intertwined and lacking modern OO features. First thing should be to create database schema and use that for a demo API. Everything else will turn out into a humongous we can't cope with (just my two cents).

#925 looks like a nice quick win before even thinking about major roadwork.

**Comment by ikedas** [comment]
> > I think that the cache mechanism of Sympa so far essentially did not make GUI response faster.
> > ```
> > * Since the most of accesses by users are sparse, cached data won't improve performance in the most of cases.  Rather, especially on the first access, performance will _never_ be improved.
> > ```
> 
> I can see two potential benefits for cached data: 1) user A initiates cache and user B benefits from it later on, but also 2) cache gets initiated at the beginning of user A session and prevents data reloading within that user session. => That's what I would like to introduce as an alternative to #925. Running a single SQL query early and populate the cache to prevent one SQL query per list, to fulfill that request.

- On 1), as I wrote "the most of accesses by users are sparse",  it is not often possible that both user A and user B are interested in the same objects (lists and so on): Rather, cache miss seems more often possible.
- On 2), the first accesses to objects cannot be previously cached and they are still slow.  After all, IMHO, users complain about the slow access, even if succeeding accesses would be cached and faster.

I think we'd be better to reduce and fasten accesses to backend (database/filesystem) rather than focusing on cache mechanism (as we cannot make "caching everything" approach).

----
Honestly, I feel Perl5 is good for prototyping but is not good for performance tuning: Most of logic may be 100s to 1000s of times faster using the other languages.
And honestly I feel Perl5 is not good for building modern GUI. 　I haven't got the slightest idea how to realize web UI like sympa-community/sympa-vue#1 with Perl5.
I got off track.  Of course I will appreciate the people wroking for any kind of improvement of wwsympa.

**Comment by salaun-urennes1** [comment]
I'm concerned about short-term gain for our production mailing list service: I wish I could get the performances we had with previous Sympa versions.

Therefore I will go on my investigation in that direction and will submit a PR as soon as possible.

**Comment by racke** [comment]
@salaun-urennes1  which versions are you referring to?

**Comment by salaun-urennes1** [comment]
@racke it's hard to say because we don't check Sympa GUI response time after each update. I'd say the web interface response times were pretty good 2 years ago. Over the last years we've received many complaints from our end-users who can't access the `/lists` page anymore.

Note that olivov related the same decrease of performances in #925.

I suppose this is related to code refactoring...

**Comment by ikedas** [comment]
That refactoring (or reconstructing) may be regarding this issue:
- Owners and editors should be stored only in database, not in config #49

It is a root of several bugs such as #7, #11. In short, these were caused by missing synchronization between three layers: config, database and "cache" on memory.

These have been fixed by #105, #275 and so on.  By these fixes, owner/editor were removed from config, and "cache" became mostly following update in database.

In fact, response of web UI has became slower by those changes.  Because, like the bugs above, previously the cache occasionally omitted updates on backend.  It's in sense that skipping updates makes response faster.  Now the code has been more or less corrected, and **response became slower in exchange of displaying correct information**.
